File: backend/app/telemetry.py
# telemetry.py - OpenTelemetry yapılandırması
# Distributed tracing için OpenTelemetry kurulumu ve Jaeger entegrasyonu

from opentelemetry import trace  # Tracing API - span oluşturma için
from opentelemetry.sdk.trace import TracerProvider  # Tracer provider - merkezi trace yönetimi
from opentelemetry.sdk.trace.export import BatchSpanProcessor  # Span'leri batch halinde gönder
from opentelemetry.sdk.resources import Resource, SERVICE_NAME  # Servis metadata
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter  # Jaeger'a gRPC ile span gönder
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor  # FastAPI otomatik instrumentation
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor  # HTTPX otomatik instrumentation
# SQLAlchemy instrumentation - şimdilik kapalı (çok fazla span yaratıyor)
# from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from app.config import settings  # Yapılandırma ayarları


def setup_telemetry():
    """
    OpenTelemetry telemetry'yi başlat
    
    Bu fonksiyon:
    1. Tracer provider oluşturur (merkezi trace yönetimi)
    2. Jaeger exporter ekler (span'leri Jaeger'a gönderir)
    3. FastAPI, HTTPX, SQLAlchemy instrumentation yapar (otomatik tracing)
    
    Not: Uygulama başlarken bir kez çağrılmalı (main.py'de)
    """
    
    # 1. Resource oluştur - servis metadata'sı
    # Bu bilgiler Jaeger'da görünecek (service adı, version, environment)
    resource = Resource.create(
        {
            SERVICE_NAME: settings.OTEL_SERVICE_NAME,  # Servis adı - "madlen-chat-backend"
            "service.version": "1.0.0",  # Servis versiyonu
            "deployment.environment": settings.ENV,  # Ortam - development/production
        }
    )
    
    # 2. Tracer Provider oluştur - merkezi trace yönetici
    # Tüm span'ler bu provider üzerinden yönetilir
    tracer_provider = TracerProvider(resource=resource)
    
    # 3. OTLP Exporter oluştur - Jaeger'a span gönderme (gRPC)
    # gRPC için endpoint: config'den oku ve insecure=True (lokalde TLS yok)
    otlp_exporter = OTLPSpanExporter(
        endpoint=settings.OTEL_EXPORTER_OTLP_ENDPOINT,
        insecure=True,
        timeout=10,
    )
    
    # 4. Batch Span Processor ekle - span'leri toplu gönder
    # Her span'i tek tek göndermek yerine batch halinde gönderir (performans için)
    span_processor = BatchSpanProcessor(otlp_exporter)
    tracer_provider.add_span_processor(span_processor)  # Provider'a processor ekle
    
    # 5. Global tracer provider'ı ayarla - uygulama genelinde kullanılsın
    trace.set_tracer_provider(tracer_provider)
    
    # 6. Otomatik instrumentation'ları ekle
    # Bu instrumentation'lar otomatik olarak span oluşturur
    
    # HTTPX instrumentation - tüm HTTP client çağrıları otomatik trace edilir
    # OpenRouter API çağrıları buradan trace edilecek
    HTTPXClientInstrumentor().instrument()
    
    # SQLAlchemy instrumentation - KAPATILDI (çok fazla span oluşturuyordu - 100+ span/request)
    # Gerekirse kritik query'ler için manuel span eklenebilir
    # SQLAlchemyInstrumentor().instrument()
    
    # Basit bir başlangıç span'i oluşturarak exporter'in çalıştığını doğrula
    tracer = trace.get_tracer("startup")
    with tracer.start_as_current_span("otel.startup.ping") as span:
        span.set_attribute("otel.endpoint", settings.OTEL_EXPORTER_OTLP_ENDPOINT)
        span.set_attribute("service.name", settings.OTEL_SERVICE_NAME)

    print("✅ OpenTelemetry başlatıldı - Jaeger'a bağlandı")  # Başarı mesajı


def get_tracer(name: str = __name__):
    """
    Tracer al - custom span'ler için
    
    Kullanım:
        tracer = get_tracer("my_service")
        with tracer.start_as_current_span("my_operation"):
            # işlemler
    
    Args:
        name: Tracer adı (genelde module adı)
        
    Returns:
        Tracer: OpenTelemetry tracer instance
    """
    return trace.get_tracer(name)  # Global tracer provider'dan tracer al

File: frontend/src/telemetry/telemetry.ts
// telemetry/telemetry.ts - OpenTelemetry yapılandırması (Frontend)
// Distributed tracing için OpenTelemetry kurulumu ve Jaeger entegrasyonu

import { WebTracerProvider } from '@opentelemetry/sdk-trace-web' // Web tracer provider
import { BatchSpanProcessor } from '@opentelemetry/sdk-trace-base' // Batch processor - performans için
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http' // Jaeger'a HTTP ile gönder
import { Resource } from '@opentelemetry/resources' // Servis metadata
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions' // Semantic conventions
// Auto instrumentation - conditional import to avoid build issues
// import { registerInstrumentations } from '@opentelemetry/instrumentation'
import { FetchInstrumentation } from '@opentelemetry/instrumentation-fetch' // Fetch API instrumentation
import { trace } from '@opentelemetry/api' // Tracing API

// OpenTelemetry'yi başlat
export function setupTelemetry() {
  // 1. Resource oluştur - servis metadata'sı (Jaeger'da görünecek)
  const resource = new Resource({
    [ATTR_SERVICE_NAME]: 'madlen-chat-frontend', // Servis adı
    'service.version': '1.0.0', // Servis versiyonu
    'deployment.environment': 'development', // Ortam
  })
  
  // 2. Tracer Provider oluştur - merkezi trace yönetici
  const provider = new WebTracerProvider({
    resource: resource, // Yukarıda oluşturduğumuz resource
  })
  
  // 3. OTLP Exporter oluştur - Jaeger'a trace gönderme
  const exporter = new OTLPTraceExporter({
    url: 'http://localhost:4318/v1/traces', // Jaeger OTLP endpoint
  })
  
  // 4. Batch Span Processor ekle - span'leri toplu gönder (performans)
  provider.addSpanProcessor(new BatchSpanProcessor(exporter))
  
  // 5. Provider'ı global olarak kaydet - uygulama genelinde kullan
  provider.register()
  
  // 6. Fetch instrumentation - API çağrılarını trace et (manuel)
  // Not: registerInstrumentations yerine direkt instrumentation kullanıyoruz
  const fetchInstrumentation = new FetchInstrumentation({
        applyCustomAttributesOnSpan: (span, request) => {
      if (request instanceof Request) {
          span.setAttribute('http.url', request.url)
      }
        },
  })
  fetchInstrumentation.setTracerProvider(provider)
  fetchInstrumentation.enable()
  
  console.log('✅ Frontend OpenTelemetry başlatıldı')
}

// Tracer al - custom span'ler için
export function getTracer(name: string = 'frontend') {
  return trace.getTracer(name) // Global tracer provider'dan tracer al
}

